rfd/0073-idp-initiated-login.md

```diff
@@ -0,0 +1,67 @@
+
+OIDC does have a feature called Third Party Initiated Login (OpenID Connect Core, Section 4) that allows an initiating third-party application to bounce to the SP which sends you to the IdP and back. This can technically be employed to support IdP-initiated login, but I don't believe that OIDC is a good candidate for this feature. This is because to make IdP-initiated login useful, you need a base platform such as a dashboard (often run by the IdP) that can be used to send the user to the SP and initiate the login. Since SAML has a commonly used mechanism for IdP-initiated login, most IdP's have a feature like this for SAML.
+
+OIDC does not have a commonly used mechanism for IdP-initiated login. This has resulted in IdP support for this feature being scarce and limited. Furthermore, the Third Party Initiated Login feature of OIDC is not well specified in the standard as it is generally used to create a middleman IdP that sits between between a true SP and IdP using a completely custom flow between an SP and an IdP which you both control. Therefore OIDC lacks the tooling and product support from IdP's to make this feature useful.
+
+For this reason, the rest of the RFD focuses purely on SAML and assumes that is the only connector we will support IdP-initiated login for.
+
+### Configuration
+
+Add a new configurable field to SAML connectors called `allow_idp_initiated` which defaults to false. If set to true, Teleport will accept the SAML assertion sent by the IdP even if it is not requested by Teleport. This allows you to configure the SAML callback in an IdP provider as usual and log in to Teleport directly from the IdP dashboard.
+
+```
+kind: saml
+version: v2
+metadata:
+  name: new_saml_connector
+spec:
+  display: MyIdP
++ allow_idp_initiated: true
+  # other fields
+```
+
+### Security Considerations
+
+Allowing IdP-initiated login flows comes with a set of security tradeoffs inherent to the use of SAML and the flow itself. In this configuration, Teleport becomes more prone to replay and CSRF attacks since Teleport cannot verify that the user started the login flow intentionally. This opens up attack vectors where an attacker can trick a legitimate user into unknowingly logging into the application with the identity of the attacker.
+
+Due to the security risks below, this feature should be opt-in using the configuration option above and not enabled by default. Since this can reduce usage friction, we should still offer the feature to those that want it.
+
+#### Specific risks
+
+- Replay Attacks: Replay attacks are possible by capturing SAML assertions and possibly reusing them. We can attempt to mitigate this in Teleport by storing a cache of SAML assertion IDs that are valid for as long the SAML assertion is itself valid. This may not always be feasible depending on the data we receive from the IdP, but it is an useful mitigation we should employ. 
```

Comment:
This is a question since I'm not personally familiar with SAML: is there any way to mitigate these replays with one time use identifiers/unique ids? I see there's an assertion for `OneTimeUse` and the docs [here](http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0-cd-02.html#4.4.2.Assertion,%20Subject,%20and%20Statement%20Structure|outline) indicate a `transient identifier`. 

I'm guessing this is prohibitive because we don't control the initiation and the external provider is only guaranteed to provide a subset of fields?

Reply:
Yep, there is indeed a transient identifier but I don't think all IdP's provide it. That doesn't prevent stealing but it does certainly help against replay. That's what the latter part of the paragraph talks about. When we see an identifier like this, we should cache it with a TTL of the assertion lifetime with a backend key to ensure it's never replayed.

Reply:
Gotcha, thanks.